### PR TITLE
Redeuce trust assumptions host <-> agreements

### PR DESCRIPTION
**Pull request by hellwolf**
## Some safety principles:

- Agreement can make mistakes.
- Host should double check these errors with runtime asserts, with
little gas costs.
- Preferring to catch these assets if any during development phases.
- Runtime assets shouldbe continuously monitored, since they could
cause non-closable termination callbacks.

## Changes:

- ISuperfluid.sol:
  - added documentations for agreement framework operations.
- Superfluid.sol:
  - validCtx -> requireValidCtx, interfacing users and apps.
  - assertValidCtx, for agreement framework operations.
  - trust assumption of appCallbackPop remains due to lack of trusted
    callback statck implementation.
- AgreementLibrary.sol
  - added security notes for implementers
  - make callAppAfterCallback
- ConstantFlowAgreementV1.sol
  - ctxUseAllowance to use correct newCtx

### CODE REVIEW

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol [hunk 1]
```diff
@@ -499,6 +501,7 @@ contract Superfluid is
             }
         } else {
             newCtx = ctx;
+            // no change to the stamped ctx, no stampin gneeded
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/superfluid/Superfluid.sol]: *small typo
fix: no stamping needed

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol [hunk 2]
```diff
@@ -511,6 +514,7 @@ contract Superfluid is
     )
         external override
         onlyAgreement
+        //assertValidCtx(ctx)
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/superfluid/Superfluid.sol]: I don't think this should be commented out unless the interface file is missing a security note for `appCallbackPush`

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol (hunk 1 above)
Comment (hellwolf) [reply packages/ethereum-contracts/contracts/superfluid/Superfluid.sol]: it was not necessary comment, something added while testing out some logic. It is deleted for now.

**Comment by 0xdavinchee** [comment]
@hellwolf did you still want to add the bad agreement case?

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol (hunk 1 above)
Comment (hellwolf) [reply packages/ethereum-contracts/contracts/superfluid/Superfluid.sol]: > @hellwolf did you still want to add the bad agreement case?

yea, gonna add it

**Comment by hellwolf** [comment]
> @hellwolf did you still want to add the bad agreement case?

It's done, please review again.

File: packages/ethereum-contracts/test/contracts/libs/CFAv1Library.test.js [hunk 3]
```diff
@@ -1,9 +1,9 @@
 const TestEnvironment = require("../../TestEnvironment");
 
-// const {BN, expectRevert} = require("@openzeppelin/test-helpers");
+// const {BN, expectRevertedWith} = require("@openzeppelin/test-helpers");
 // const {web3tx, toWad, toBN} = require("@decentral.ee/web3-helpers");
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/test/contracts/libs/CFAv1Library.test.js]: nbd, but good to just remove commented out imports

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol [hunk 4]
```diff
@@ -301,8 +325,19 @@ interface ISuperfluid {
...
 
+    /**
+     * @dev (For agreements) Pop from the current app callback stack
+     * @param  ctx                     The ctx that was pushed before the callback stack.
+     * @param  appAllowanceUsedDelta   App allowance used by the app.
+     * @return newCtx
+     *
+     * [SECURITY] NOTE:
+     * - Here we cannot do assertValidCtx(ctx), since we do not really save the stack in memory.
+     * - Hence there is still implicit trust that the agreement library doing the callback push/pop pair correctly.
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol]: another minor comment:
`that the agreement library doing the callback push/pop pair correctly.` => `that the agreement library **is** doing the callback push/pop pair correctly.`

File: packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.js [hunk 5]
```diff
@@ -619,269 +622,397 @@ describe("Superfluid Host Contract", function () {
...
-                    "revert"
-                );
-                await expectRevert(
-                    superfluid.ctxUseAllowance("0x", 0, 0),
-                    "revert"
-                );
+                it("#6.2 use agreement framework as an unregisterred agreement", async () => {
+                    const reason = "SF: sender is not listed agreeement";
 
-                // call from an unregisterred mock agreement
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.js:663]: unregistered*

**Review by 0xdavinchee — APPROVED** [review]

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol (hunk 2 above)
Comment (hellwolf) [reply packages/ethereum-contracts/contracts/superfluid/Superfluid.sol]: it's resolved

File: packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.js (hunk 5 above)
Comment (hellwolf) [reply packages/ethereum-contracts/test/contracts/superfluid/Superfluid.test.js:663]: fixed

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol (hunk 4 above)
Comment (hellwolf) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol]: fixed

**Pull request merged**